clippy_lints/src/manual_is_ascii_check.rs

```diff
@@ -112,19 +122,58 @@ impl<'tcx> LateLintPass<'tcx> for ManualIsAsciiCheck {
                 limits: RangeLimits::Closed,
             }) = higher::Range::hir(receiver)
         {
-            let range = check_range(start, end);
-            if let ExprKind::AddrOf(BorrowKind::Ref, _, e) = arg.kind {
-                check_is_ascii(cx, expr.span, e, &range);
-            } else {
-                check_is_ascii(cx, expr.span, arg, &range);
+            let arg_ty = cx.typeck_results().expr_ty(arg).peel_refs();
+            if matches!(arg_ty.kind(), ty::Param(_)) {
+                return;
             }
+
+            let arg = peel_ref_operators(cx, arg);
+            let ty_sugg = get_ty_sugg(arg, &self.parent_closure_params_map, start);
+            let range = check_range(start, end);
+            check_is_ascii(cx, expr.span, peel_ref_operators(cx, arg), &range, ty_sugg);
+        }
+    }
+
+    fn check_expr_post(&mut self, _: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let ExprKind::Closure(_) = expr.kind {
+            self.parent_closure_params_map = FxHashMap::default();
         }
     }
 
     extract_msrv_attr!(LateContext);
 }
 
-fn check_is_ascii(cx: &LateContext<'_>, span: Span, recv: &Expr<'_>, range: &CharRange) {
+fn get_ty_sugg(arg: &Expr<'_>, map: &FxHashMap<HirId, Span>, range_expr: &Expr<'_>) -> Option<(Span, &'static str)> {
+    let ExprKind::Lit(lit) = range_expr.kind else {
+        return None;
+    };
+    let sugg_ty_span = path_to_local(arg).and_then(|id| map.get(&id)).copied()?;
```

Comment:
If it works it would be good to use `find_binding_init` here rather than storing a map, would let us avoid any issues with nested closures

Reply:
Unfortunately I couldn't get it to work...

but regarding nested closures, I now clear the map only when entering new modules rather than leaving another closure, which should be able to prevent it.